src/MobileDevice/widget/MobileDevice.ts

```diff
@@ -0,0 +1,151 @@
+
+type onDeviceReadyActions = "doNothing" | "showPage" | "callMicroflow";
+
+class MobileDevice extends WidgetBase {
+    private deviceIdAttribute: string;
+    private deviceTypeAttribute: string;
+    private appVersionVersionAttribute: string;
+    private appVersionBuildAttribute: string;
+    private microflow: string;
+    private onDeviceReadyAction: onDeviceReadyActions;
+    private page: string;
+    private mxObject: mendix.lib.MxObject;
+    private onNavigateBack: boolean;
+    // internal variables
+    private deviceReadyEvent: dojoEvent | null;
+
+    postCreate() {
+        this.showError(this.validateProps());
+        this.setUpWidgetDom();
+    }
+
+    update(mxObject: mendix.lib.MxObject, callback: () => void) {
+        this.mxObject = mxObject;
+        if (!window.device) {
+            mxObject.set(this.deviceIdAttribute, "");
+            mxObject.set(this.deviceTypeAttribute, "Web");
+            this.commit(mxObject);
+        }
+        this.setUpEvents();
+
+        callback();
+    }
+
+    uninitialize(): boolean {
+        this.removeEvents();
+
+        return true;
+    }
+
+    private showError(message: string) {
+        if (message) {
+            domConstruct.create("div", {
+                class: "alert alert-danger",
+                innerHTML: message
+            }, this.domNode);
+        }
+    }
+
+    private setUpEvents() {
+        if (window.device) {
+            this.deviceReadyEvent = dojoOn(document, "deviceReady", dojoLang.hitch(this, this.onDeviceReady));
+        }
+    }
+
+    private removeEvents() {
+        if (window.device) {
+            if (this.deviceReadyEvent) {
+                this.deviceReadyEvent.remove();
+                this.deviceReadyEvent = null;
+            }
+        }
+    }
+
+    private setUpWidgetDom() {
+        domConstruct.create("div", {
+            class: "mobile-device-widget",
+            id: "mobile-device"
+        }, document.body, "first");
+    }
+
+    private onDeviceReady() {
+        if (cordova.hasOwnProperty("getAppVersion")) {
+            cordova.getAppVersion.getVersionNumber((versionNumber: string) => {
+                if (versionNumber) {
+                    cordova.getAppVersion.getVersionCode((versionCode: string) => {
+                        if (versionCode) {
+                            this.mxObject.set(this.deviceIdAttribute, window.device.uuid);
+                            this.mxObject.set(this.deviceTypeAttribute, window.device.platform);
+                            this.mxObject.set(this.appVersionBuildAttribute, versionCode);
+                            this.mxObject.set(this.appVersionVersionAttribute, versionNumber);
+                            this.commit(this.mxObject);
+                        }
+                    });
+                }
+            });
+        } else {
+            window.mx.ui.error("Mobile device widget: add cordova-plugin-app-version to your project config");
```

Comment:
It is true that in the new hybrid application generator the application version plugin is always there. The question is what if the developer edits and removes it? However i am asking him for comfirmation